### PR TITLE
fix(pages): synchronize generated form values with mounted controls

### LINKED ISSUE #6268: bug(pages): form! initial values and DOM synchronization are incomplete
## Prerequisites

- [x] I have searched open and closed issues for initial values, controlled inputs, and form reset.
- [x] I reviewed #4386, #5258, #6212, #6221, and #6240. The remaining branch and widget coverage gaps are distinguished below.

## Bug Description

On current `main`, ordinary `form!` fields hold the correct initial Rust value but render browser defaults. Browser input updates Rust state, while programmatic signal updates and `use_form(&form).build().reset()` leave the mounted controls unchanged.

This is a rendering/binding gap, not a failure to initialize the Rust signals. It also affects hidden inputs and native SSR output.

## Steps to Reproduce

1. Build a minimal Pages fixture from the revision below with this form:

```rust
use reinhardt_pages::{form, use_form};

let form = form! {
    name: ValueProbe,
    action: "/test",
    fields: {
        name: CharField { label: "Name", initial: "Ada" }
        subscribed: BooleanField { label: "Subscribe", initial: true }
        status: ChoiceField<String> {
            label: "Status", initial: "review",
            choices: [("open", "Open"), ("review", "In review")],
        }
        token: HiddenField<String> { initial: "fixture-hidden" }
        color: CharField {
            label: "Color", widget: ColorInput, initial: "#336699",
        }
        level: IntegerField {
            label: "Level", widget: RangeInput, initial: 25,
            min: 0, max: 100,
        }
    }
};
let runtime = use_form(&form).build();
assert_eq!(form.name().get(), "Ada");
let page = form.clone().into_page();
```

2. Mount `page` normally, retaining `form` and `runtime`. Inspect native SSR separately with `form.clone().into_page().render_to_string()`.
3. Compare the mounted properties with the declared values:

| Control | Expected | Observed on main |
|---|---|---|
| `#name.value` | `"Ada"` | `""` |
| `#subscribed.checked` | `true` | `false` |
| `#status.value` | `"review"` | `"open"` |
| `#token.value` | `"fixture-hidden"` | `""` |
| `#color.value` | `"#336699"` | `"#000000"` |
| `#level.value` | `"25"` | `"50"` |

4. Type `"Before update"` in the name input, call `form.name().set("Grace".to_owned())`, and assert both the Rust value and `input.value` equal `"Grace"`. Only the Rust assertion passes.
5. Type `"Before reset"`, call `runtime.reset()`, and assert both values equal `"Ada"`. Only the Rust assertion passes.

## Expected Behavior

Generated controls render their declared state in HTML and on initial mount. Bound controls synchronize both directions; runtime reset restores their displayed state using the existing mounted nodes. File inputs retain native restrictions on programmatic values.

## Actual Behavior

The complete fixture covered 30 fields across all 21 field types and 21 accepted built-in widgets:

- All 30 initial Rust values matched; 25 initial DOM values did not.
- All 27 DOM-to-Rust interaction assertions passed.
- Five programmatic updates and five resets changed Rust values but failed to update the corresponding text, checkbox, select, multi-select, and number controls.
- SSR omitted the corresponding initial value/checked/selected state.
- Empty text, unchecked boxes, the wrong selected option, a black color swatch, and a centered slider were visible at both viewport widths.

## Environment

- Reproduced revision: `main@0471a9d49864b871bc2365a844d88634f03848e6`; fetched `origin/main` matched this revision on 2026-09-07.
- Crate: `reinhardt-pages 0.3.15`, built from workspace source.
- Rust: 1.96.0; native and `wasm32-unknown-unknown` targets.
- OS: macOS 26.6.2; Chrome 152.0.7977.77 via Playwright.
- Browser viewports: 1280×1024 and 390×844, DPR 1, native browser styles.
- The freshly built WASM file and browser-fetched bytes had the same SHA-256: `e4322cfc21ecf5fc30b3c0cee6e1815483a9b7e88737b816bb225083ec0c5751`.
- No console errors, page errors, failed requests, or HTTP error responses were observed in the assertion run.

### Root cause and affected paths

1. [`generate_field_initializers`](https://github.com/kent8192/reinhardt-web/blob/0471a9d49864b871bc2365a844d88634f03848e6/crates/reinhardt-pages/macros/src/form/codegen.rs#L2281) initializes the signals correctly; this differs from the old signal-initialization bug #4386.
2. [`generate_into_page`](https://github.com/kent8192/reinhardt-web/blob/0471a9d49864b871bc2365a844d88634f03848e6/crates/reinhardt-pages/macros/src/form/codegen.rs#L3121) clones bound signals. Top-level fields and nested `FieldGroup` fields both reach [`generate_field_view`](https://github.com/kent8192/reinhardt-web/blob/0471a9d49864b871bc2365a844d88634f03848e6/crates/reinhardt-pages/macros/src/form/codegen.rs#L3543).
3. [The shared renderer](https://github.com/kent8192/reinhardt-web/blob/0471a9d49864b871bc2365a844d88634f03848e6/crates/reinhardt-pages/macros/src/form/codegen.rs#L4386) emits controls without initial `value`, textarea content, `checked`, or selected-option state. Dynamic `RadioSelect` is a separate path that already reads its signal for checked state.
4. [`generate_bind_listener`](https://github.com/kent8192/reinhardt-web/blob/0471a9d49864b871bc2365a844d88634f03848e6/crates/reinhardt-pages/macros/src/form/codegen.rs#L4994) only emits DOM event listeners. Despite the two-way-binding comment, this main-branch path installs no state-to-control subscription.
5. [Collection fields use a separate renderer](https://github.com/kent8192/reinhardt-web/blob/0471a9d49864b871bc2365a844d88634f03848e6/crates/reinhardt-pages/macros/src/form/codegen.rs#L3945), including some existing value/checked serialization. Do not assume they have the same behavior as ordinary fields without testing.

### Existing work and repair scope

#6221 was closed by #6240, merged into `develop/0.4.0` on 2026-09-06; it is not present in the tested main revision. Reuse and reconcile that implementation instead of reimplementing the same binding machinery.

Source inspection of [`generate_generated_control_binding` on develop](https://github.com/kent8192/reinhardt-web/blob/e89a852818f607bc9547b03f124c2109ba6be9c6/crates/reinhardt-pages/macros/src/form/codegen.rs#L7642) shows a narrower widget/type match with an empty fallback. It still omits categories exercised here, including hidden, range, color, telephone, month/week and typed temporal controls, decimal, UUID/IP, and JSON. This branch comparison is source analysis; the recorded browser results above are for main only. A blanket claim that #6240 resolves the complete widget matrix is therefore insufficient.

Repair the supported stable form contract conservatively on main and preserve the existing develop binding/reset work when integrating it. Audit the ordinary, grouped, collection, and model rendering paths; do not pull unrelated 0.4 APIs into a stable repair.

### Acceptance criteria

- [ ] Exact native-HTML and WASM DOM assertions cover initial values for every supported field/control category, with explicit file-input exceptions.
- [ ] Textarea, checkbox, single/multiple select, hidden, temporal, color, range, and structured value formatting all have meaningful assertions.
- [ ] Programmatic updates and runtime reset synchronize mounted controls; tests first establish a different DOM value so a no-op cannot pass.
- [ ] Ordinary edits and reset preserve DOM node identity, focus behavior, and existing hydration/reset rules.
- [ ] `bind: false` has an explicit initial-render contract and no unintended ongoing binding.
- [ ] Desktop and narrow-viewport screenshots verify the corrected populated/checked/selected states against an independently specified reference.
- [ ] Regression coverage and documentation describe the supported value conversions and any intentionally unsupported combinations.

🤖 Generated with [Codex](https://openai.com/codex/)

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

Generated `form!` fields can hold the correct Rust value while displaying browser defaults, and signal updates or runtime reset can leave the controls stale. Render the declared values in native HTML and retain control bindings so mounted values follow the signals without replacing the controls.

- Share typed serialization and bindings across ordinary fields, groups, and collections; evaluate each static option expression once for its value and selection.
- Preserve hydration edits and browser-selected files, field/path-specific runtime overrides and reconciled pristine defaults, source order for unchanged multiple selections, focus, and control identity.
- Preserve leading textarea newlines, treat CR/CRLF parser normalization as unchanged hydration, and retain controlled select state through selective hydration; synchronize options when they arrive or are replaced. Validate unbound textarea snapshots as parser-normalized text, retaining empty and whitespace-only defaults.
- Publish native reset values in one reactive batch and clear touched/error metadata without edit or validation notifications, including unchanged-value resets. Retain custom-widget error subscriptions, prioritize layout work, and deduplicate chained notifications until the callback queue is drained.
- Preserve existing collection row keys and path subscriptions during same-length reset, and document `bind: false` snapshots and native file-input restrictions.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (code change that neither fixes a bug nor adds a feature)
- [x] Documentation update
- [ ] Performance improvement
- [ ] Code quality improvements
- [ ] CI/CD changes
- [ ] Other (please describe):

## Breaking Change Assessment

- [x] **No** - This change is backward compatible
- [ ] **Yes** - This change breaks existing public API or behavior

## Motivation and Context

At `main@0471a9d49864b871bc2365a844d88634f03848e6`, the issue's fixture has 30 correct initial Rust values but 25 incorrect initial DOM values. `crates/reinhardt-pages/macros/src/form/codegen.rs` emits event listeners without a retained signal-to-control subscription and omits native initial value, checked, or selected state.

The stable repair reconciles the control-descriptor approach from #6240 with the complete supported widget matrix. Core carries the descriptor; Pages owns its controller through mount, hydration, and disposal. The stable branch has no ModelForm rendering API, so this repair does not import the development branch's 0.4 APIs.

Fixes #6268

Related to: #6240

## How Was This Tested?

- The affected native Pages integration run passed 70 tests: initial values/collections (34), form runtime (31), attributes (3), and radio groups (2). The final reconciliation unit run passed all 18 tests.
- The final Core reactive run passed 59 tests, including nested batches, timing priority, and deduplicated chained layout writes. The expanded textarea test passed 11 newline/escaping cases in both all-feature and focused builds. Earlier in this repair, the full affected Core `reactive,types,exception,page` suite passed 663 tests.
- All 84 form code-generation tests passed. A changing-expression regression proves static choice values are evaluated once across ordinary, grouped/optgroup, and collection single/multiple selects.
- The final Chrome 152 WASM run passed 32 tests: value synchronization (20), hydration edge cases (5), attributes (3), and radio groups (4). Assertions cover field/path overrides, refreshed scalar and collection defaults before hydration alongside unrelated browser edits, actual File identity, selection order, LF/CR/CRLF parsing, unbound empty/whitespace-only textarea snapshots and rejection of changed text, selective option traversal, native reset metadata/events, cancellation, focus, and retained nodes.
- Native Core/Pages/macros and WASM Pages Clippy (including the changed integration targets) passed with warnings, TODO, unimplemented, and dbg lints denied. Core's page-only WASM feature check passed earlier in the repair.
- All-feature Core/Pages/macros rustdoc passed with warnings denied. Repository formatting checked 3,013 files with zero errors; all changed Rust files passed the final formatting and diff checks. actionlint passed.

```sh
cargo test -p reinhardt-pages --features chrono,uuid,testing --test form_value_sync --test use_form_integration --test form_field_attributes --test form_radio_group
cargo test -p reinhardt-core --lib --no-default-features --features reactive,types,exception,page
cargo test -p reinhardt-pages-macros --lib form::codegen::tests
cargo clippy -p reinhardt-pages --target wasm32-unknown-unknown --features chrono,uuid,testing --lib --test form_value_sync --test form_control_hydration_regressions -- -D warnings
cargo check -p reinhardt-core --no-default-features --features page --target wasm32-unknown-unknown
cargo doc -p reinhardt-core -p reinhardt-pages -p reinhardt-pages-macros --all-features --no-deps
cargo make fmt-check
```

Browser tests used wasm-bindgen-test-runner 0.2.128 and ChromeDriver 152. The integration targets are included in the WASM CI workflow. Full workspace tests, the complete WASM suite, other browsers, and real devices were not run for this repair. Remote CI is tracked separately.

### Before

The reproduced main fixture shows empty text, unchecked boxes, incorrect selections, a black color swatch, and a centered range input, as recorded in #6268.

### After

Desktop (1280×1024) and narrow (390×844) captures were compared with an independently specified native HTML reference. All 30 initial values matched per viewport; all 60 control crops had zero Pixelmatch differences at threshold 0.1 with anti-aliasing excluded. Tiny raw-channel differences remain, so the captures are not bit-identical. Both viewports passed 27 edits, signal restoration, and runtime reset while preserving 39 control/option nodes and input focus. Screenshot files and the comparison report are retained in the task deliverables.

The repository fixture is `crates/reinhardt-pages/tests/fixtures/form_value_sync.rs`; behavioral regressions are in `crates/reinhardt-pages/tests/form_value_sync.rs`.

## Checklist

- [x] I have followed the [Contributing Guidelines](../blob/main/CONTRIBUTING.md)
- [x] I have followed the [Commit Guidelines](../blob/main/instructions/COMMIT_GUIDELINE.md)
- [x] I have updated the documentation (if applicable)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have tested with all affected database backends (if applicable)
- [ ] I have formatted the code with `cargo make fmt-fix`
- [ ] I have checked the code with `cargo make clippy-check`
- [ ] I use self-hosted runner for CI (Repository owner only)

The test checkbox refers to the affected suites listed above. Database backends are unaffected. Formatting was verified with `cargo make fmt-check`; Clippy used the explicit affected-package and target commands above.

## Related Issues

- #6268
- #6240

### Type Label (select one)

- [x] `bug` - Bug fix
- [ ] `enhancement` - New feature or improvement
- [ ] `documentation` - Documentation update
- [ ] `performance` - Performance improvement
- [ ] `refactoring` - Code refactoring
- [ ] `code-quality` - Code quality improvements

### Additional Labels (apply alongside type label when applicable)

- [ ] `breaking-change` - Breaking change (**MUST** match "Yes" in Breaking Change Assessment above)

### Scope Label (select all that apply)

- [ ] `database` - Database layer, schema, migrations
- [ ] `auth` - Authentication, authorization, sessions
- [ ] `orm` - ORM layer, models, query builder
- [ ] `http` - HTTP layer, handlers, middleware
- [ ] `routing` - URL routing, path matching
- [ ] `api` - REST API, serializers, views
- [ ] `admin` - Admin interface, admin panels
- [x] `forms` - Form handling, validation, rendering
- [ ] `graphql` - GraphQL schema, resolvers
- [ ] `websockets` - WebSocket connections, handlers
- [ ] `i18n` - Internationalization, localization
- [ ] `ci-cd` - CI/CD workflow changes

### Priority Label (for maintainers)

- [ ] `critical` - Blocks release or major functionality
- [ ] `high` - Important fix or feature
- [ ] `medium` - Normal priority
- [ ] `low` - Minor fix or enhancement

---

**Additional Context:**

Structural collection changes retain the stable renderer's replacement behavior. Files remain browser-owned: generated controls never assign nonempty file values. Existing unsupported widget combinations remain rejected by the macro validator.

🤖 Generated with [Codex](https://openai.com/codex/)

### CODE REVIEW

File: crates/reinhardt-pages/macros/src/form/codegen.rs [hunk 1]
```diff
@@ -4097,42 +4140,74 @@ fn generate_collection_field_view(
...
+		return TokenStream::new();
+	}
+	let field_name = &field.name;
+	let field_type = field_type_to_value_type(&field.field_type);
+	let assignment = collection_field_assignment(field, pages_crate, collection_name);
+	let setup = quote! {
+		let __read_value = {
+			let collection = __collection_signal.clone();
+			let key = item.key();
+			// ponytail: linear lookup per field; index keys if large arrays become common.
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/macros/src/form/codegen.rs]: Clarified in `4ebd85e2fb`. The comment in `crates/reinhardt-pages/macros/src/form/codegen.rs` describes the implemented lookup's O(rows) cost and its scaling boundary. It does not mark missing functionality: stable collection keys are resolved correctly through the existing row vector. The explicit `ponytail:` cost/tradeoff note is retained; adding another index is unnecessary for this repair. Collection update, hydration, reset, and identity regressions pass.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/macros/src/form/codegen.rs [hunk 2]
```diff
@@ -4485,6 +4685,8 @@ fn generate_field_view(
 						#native_attrs
 						#custom_attrs
 						#event_listener
+					#control_binding
+					.child(#initial_value)
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/macros/src/form/codegen.rs:4743]: Fixed in `4ebd85e2fb`. The shared serializer in `crates/reinhardt-core/src/types/page.rs` compensates for the leading textarea line feed discarded by HTML parsing, while direct mounting keeps the original text. Native and Chrome regressions cover empty text, leading newlines, escaping, hydration, and native reset.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/macros/src/form/codegen.rs [hunk 3]
```diff
@@ -1598,6 +1574,7 @@ fn generate_form_runtime_contract(
 			}
 
 			fn runtime_apply_field_value(&self, field: Self::Field, values: &Self::Values) {
+				self.__explicitly_reset.set(true);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/macros/src/form/codegen.rs]: Fixed in `4ebd85e2fb`. `runtime_apply_field_value` in `crates/reinhardt-pages/macros/src/form/codegen.rs` now marks only that field as source-preferred during hydration. The Chrome regression resets one field before attachment and verifies an unrelated browser edit and all control nodes are preserved.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/src/dom/control_binding.rs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+					.dyn_ref::<web_sys::HtmlTextAreaElement>()
+					.map(web_sys::HtmlTextAreaElement::value)
+			})
+			.map(ControlValue::Text),
+		ControlKind::Checkbox | ControlKind::Radio => element
+			.dyn_ref::<web_sys::HtmlInputElement>()
+			.map(|input| ControlValue::Checked(input.checked())),
+		ControlKind::File => element.dyn_ref::<web_sys::HtmlInputElement>().map(|input| {
+			ControlValue::Checked(input.files().is_some_and(|files| files.length() > 0))
+		}),
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/src/dom/control_binding.rs]: Fixed in `4ebd85e2fb`. `crates/reinhardt-pages/src/dom/control_binding.rs` snapshots the actual browser-selected File and the generated binding adopts it into the field signal. Programmatic DOM writes remain clear-only. Chrome tests verify the exact selected File objects survive hydration for both file and image fields without replacing inputs.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/macros/src/form/codegen.rs [hunk 5]
```diff
@@ -1585,7 +1559,9 @@ fn generate_form_runtime_contract(
 			}
 
 			fn runtime_apply_values(&self, values: &Self::Values) {
+				self.__explicitly_reset.set(true);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/macros/src/form/codegen.rs:1570]: Fixed in `4ebd85e2fb`. Generated scalar and collection-path runtime setters record source preference for the affected field or stable item key in `crates/reinhardt-pages/macros/src/form/codegen.rs`. Chrome tests construct the Page first, apply explicit setters, then hydrate stale DOM while preserving unrelated browser edits and node identity.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/src/dom/control_binding.rs [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+			(read_control(element, binding.kind()), prefer_source)
+		})
+		.collect::<Vec<_>>();
+	batch(|| {
+		for ((_, binding), (snapshot, prefer_source)) in controls.iter().zip(&snapshots) {
+			if !prefer_source
+				&& let Some(snapshot) = snapshot
+				&& *snapshot != binding.read_untracked()
+			{
+				binding.write(snapshot.clone());
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/src/dom/control_binding.rs]: Fixed in `4ebd85e2fb`. Hydration in `crates/reinhardt-pages/src/dom/control_binding.rs` compares multiple selections by membership before adopting the DOM snapshot. An unchanged selection retains the source vector order and pristine runtime state; the existing edited-selection regression still verifies DOM adoption.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/src/hydration/reconcile.rs [hunk 7]
```diff
@@ -301,20 +302,43 @@ fn reconcile_element_at_path(
...
+		return Ok(());
+	}
+	let controlled_select = controlled_select
+		|| el_view.bound_control().is_some_and(|binding| {
+			matches!(
+				binding.kind(),
+				crate::component::ControlKind::SelectOne
+					| crate::component::ControlKind::SelectMany
+			)
+		});
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/src/hydration/reconcile.rs]: Fixed in `4ebd85e2fb`. `crates/reinhardt-pages/src/hydration/reconcile.rs` now carries ancestor controlled-select state through options-aware traversal and island boundaries. Chrome regressions cover single/multiple select resets, an island inside an optgroup, and continued mismatch detection for unbound selected attributes.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/src/dom/control_binding.rs [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+								})
+								.collect::<Vec<_>>()
+						});
+						batch(|| {
+							for (control, value) in snapshots {
+								if control.active.get() && value != control.binding.read_untracked()
+								{
+									control.binding.write(value);
+								}
+							}
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/src/dom/control_binding.rs:229]: Fixed in `4ebd85e2fb`. Bound controls notify a form-local reset generation after all browser defaults have been adopted. `crates/reinhardt-pages/src/form_state.rs` then clears scalar, collection, and path touched/error metadata without edit or validation events, including when values already equal their defaults. Chrome regressions verify both scalar and collection runtime state and path watchers.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-core/src/reactive/runtime.rs [hunk 9]
```diff
@@ -260,9 +261,14 @@ impl Runtime {
...
 
-			// Execute layout effects synchronously
+			// Explicit batches must not expose intermediate values to layout effects.
+			let batched = *self.batch_depth.borrow() > 0;
 			for effect_id in layout_effects {
-				super::effect::Effect::execute_effect(effect_id);
+				if batched {
+					self.schedule_update(effect_id);
+				} else {
+					super::effect::Effect::execute_effect(effect_id);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-core/src/reactive/runtime.rs:270]: Fixed in `ed5a4c74a6`, included in pushed head `374275738e86e53ff300c46a6bddadec86fe6f5e`. `crates/reinhardt-core/src/reactive/effect.rs` keeps the batch active while draining callbacks and retains not-yet-run IDs in the live queue, so nested writes deduplicate against pending layout consumers. New layout work also runs before remaining passive consumers. The regression originally observed the consumer twice, including an incomplete producer snapshot; it now observes the completed snapshot once. All 59 reactive tests and the 30 affected Chrome tests passed.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/macros/src/form/codegen.rs [hunk 10]
```diff
@@ -3828,15 +3852,21 @@ fn generate_static_choice_item_view(choice: &TypedChoiceItem) -> TokenStream {
...
-fn generate_static_choice_option_view(option: &TypedChoiceOption) -> TokenStream {
+fn generate_static_choice_option_view(
+	option: &TypedChoiceOption,
+	selected: Option<&TokenStream>,
+) -> TokenStream {
 	let value = &option.value;
 	let label = &option.label;
 	let disabled = option.disabled;
 
+	let selected = selected.map(|values| quote! { .bool_attr("selected", (#values).iter().any(|selected| selected == &(#value).to_string())) });
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/macros/src/form/codegen.rs]: Fixed in `374275738e86e53ff300c46a6bddadec86fe6f5e`. `generate_static_choice_option_view` in `crates/reinhardt-pages/macros/src/form/codegen.rs` serializes the choice expression once, computes selection using that cached string, and moves the same string into the value attribute. Native and Chrome regressions use six changing expressions across ordinary, grouped/optgroup, and collection single/multiple selects, asserting exactly six evaluations and the expected selected values. All 84 code-generation tests passed.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/src/dom/control_binding.rs [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+						ControlValue::SelectedValues(browser),
+						ControlValue::SelectedValues(source),
+					) => {
+						browser.iter().all(|value| source.contains(value))
+							&& source.iter().all(|value| browser.contains(value))
+					}
+					_ => *snapshot == source,
+				};
+				if !matches_source {
+					binding.write(snapshot.clone());
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/src/dom/control_binding.rs:120]: Fixed in `a35a1c247d`, included in pushed head `374275738e86e53ff300c46a6bddadec86fe6f5e`. Textarea hydration in `crates/reinhardt-pages/src/dom/control_binding.rs` compares CRLF and bare CR as LF while preserving the original source when no edit occurred. `crates/reinhardt-core/src/types/page.rs` also compensates for leading CR/CRLF during SSR. The Chrome regression covers seven newline patterns with untouched and edited browser values, checking source/default/watch values, field/form dirty and touched state, notifications, and node identity. The expanded native textarea test and all 30 affected Chrome tests passed.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/macros/src/form/codegen.rs [hunk 12]
```diff
@@ -1739,6 +1722,35 @@ fn generate_form_runtime_contract(
...
+	quote! {
+		{
+			let __previous = self.#name.get_untracked();
+			let __values = &#values;
+			let __same_length = __previous.len() == __values.len();
+			if !__same_length {
+				self.__path_signals.borrow_mut()
+					.retain(|path, _| !path.starts_with(#path_prefix));
+			}
+			self.#name.set(__values.iter().cloned().enumerate().map(|(index, value)| {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/macros/src/form/codegen.rs:1748]: Fixed in 0438ead88004b4c5ae0bbab39dcc3221c1352ee2. `crates/reinhardt-pages/macros/src/form/codegen.rs` now marks the scalar fields and keyed collection paths actually updated by `reconcile_defaults` as source-preferred before hydration. Unchanged defaults do not acquire this preference, and dirty fields remain eligible to adopt subsequent browser edits.

The new Chrome regression in `crates/reinhardt-pages/tests/form_value_sync.rs` hydrates stale SSR after refreshing scalar and collection defaults. It verifies the refreshed source/watch/DOM values, stable row and node identity, pristine collection state, suppressed edit events, and preservation of an unrelated dirty browser control. All 70 affected native Pages tests, 84 code-generation tests, and 31 Chrome tests passed, along with native/WASM Clippy, rustdoc with warnings denied, formatting, and diff checks.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/macros/src/form/codegen.rs [hunk 13]
```diff
@@ -4154,32 +4257,210 @@ fn collection_field_checked_expr(field: &TypedFormFieldDef) -> TokenStream {
...
+	quote! {
+		.control_binding({
+			use #pages_crate::component::{ControlBinding, ControlKind, ControlValue};
+			#setup
+			ControlBinding::from_parts(ControlKind::#kind, move || #read, move |__value| {
+				#write
+			}).prefer_source_on_hydration({
+				let __explicitly_reset = __explicitly_reset.clone();
+				let __source_preferred_fields = __source_preferred_fields.clone();
+				move || __explicitly_reset.get() || __source_preferred_fields.borrow().contains(&__source_preference_key)
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/macros/src/form/codegen.rs:4400]: This is covered by 0438ead88004b4c5ae0bbab39dcc3221c1352ee2. The collection-aware `runtime_apply_pristine_values` implementation in `crates/reinhardt-pages/macros/src/form/codegen.rs` now records source preference before assigning each changed pristine scalar, as well as changed collection paths. Unchanged values do not gain preference.

The Chrome regression `pre_hydration_reconciled_collection_defaults_preserve_other_dirty_controls` in `crates/reinhardt-pages/tests/form_value_sync.rs` exercises both pristine scalar defaults and an unrelated dirty scalar beside a FieldArray. It verifies source, watch, DOM, dirty state, event counts, and retained nodes. All 31 affected Chrome tests and 70 native Pages tests pass.

🤖 Generated with [Codex](https://openai.com/codex/)

File: crates/reinhardt-pages/src/hydration/reconcile.rs [hunk 14]
```diff
@@ -301,20 +307,46 @@ fn reconcile_element_at_path(
 		});
 	}
 
-	reconcile_attrs_at_path(element, el_view, element_path.clone())?;
-	reconcile_children_at_path(element, el_view.child_views(), element_path)
+	reconcile_attrs_at_path(element, el_view, element_path.clone(), controlled_select)?;
+	if el_view.bound_control().is_some() && expected_tag == "textarea" {
+		// Textarea contents are the browser's reset value, not its live state.
+		return Ok(());
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-pages/src/hydration/reconcile.rs:313]: Fixed the unbound textarea snapshot path in 9f8c945f41fe88a5c9e7e829f7478b45606c55bc. The browser regression exposed missing-child errors for empty/newline-only defaults because normal child filtering discards whitespace text. `crates/reinhardt-pages/src/hydration/reconcile.rs` now compares textarea snapshot text with HTML CR/CRLF normalization, retaining significant whitespace and detecting actual content changes. Options-aware traversal also avoids treating raw textarea text as nested elements.

`unbound_textarea_snapshots_reconcile_parser_normalized_newlines` in `crates/reinhardt-pages/tests/form_control_hydration_regressions.rs` covers empty, CR, CRLF, leading-newline, and whitespace-only defaults through normal and strict options-aware reconciliation. It verifies unchanged snapshots after source writes, node identity, and rejection of changed text. All 32 affected Chrome tests and 18 reconciliation unit tests passed, with native/WASM Clippy, warnings-denied rustdoc, formatting, and diff checks.

🤖 Generated with [Codex](https://openai.com/codex/)
